test_util/src/pty.rs

```diff
@@ -0,0 +1,437 @@
+          application_str.as_mut_ptr(),
+          command_str.as_mut_ptr(),
+          ptr::null_mut(),
+          ptr::null_mut(),
+          FALSE,
+          EXTENDED_STARTUPINFO_PRESENT | CREATE_UNICODE_ENVIRONMENT,
+          environment_vars
+            .as_mut()
+            .map(|v| v.as_mut_ptr() as LPVOID)
+            .unwrap_or(ptr::null_mut()),
+          cwd.as_mut_ptr(),
+          &mut startup_info.StartupInfo,
+          &mut proc_info,
+        ));
+
+        // close the handles that the pseudoconsole now has
+        drop(stdin_read_handle);
+        drop(stdout_write_handle);
+
+        // start a thread that will close the pseudoconsole on process exit
+        let thread_handle = WinHandle::new(proc_info.hThread);
+        std::thread::spawn({
+          let thread_handle = thread_handle.duplicate();
+          let console_handle = WinHandle::new(console_handle);
+          move || {
+            WaitForSingleObject(thread_handle.as_raw(), INFINITE);
+            // wait for the reading thread to catch up
+            std::thread::sleep(Duration::from_millis(200));
+            // close the console handle which will close the
+            // stdout pipe for the reader
+            ClosePseudoConsole(console_handle.take());
+          }
+        });
+
+        Self {
+          stdin_write_handle,
+          stdout_read_handle,
+          _process_handle: WinHandle::new(proc_info.hProcess),
+          _thread_handle: thread_handle,
+          _attribute_list: attribute_list,
+        }
+      }
+    }
+  }
+
+  impl Read for WinPseudoConsole {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+      unsafe {
+        let mut bytes_read = 0;
+        // don't bother checking the result of this call as it will be false
+        // when the stdout pipe is closed and bytes_read will be 0 in that case
+        ReadFile(
+          self.stdout_read_handle.as_raw(),
+          buf.as_mut_ptr() as _,
+          buf.len() as u32,
+          &mut bytes_read,
+          ptr::null_mut(),
+        );
+        Ok(bytes_read as usize)
+      }
+    }
+  }
+
+  impl Pty for WinPseudoConsole {
+    fn write_text(&mut self, text: &str) {
+      // windows psuedo console requires a \r\n to do a newline
+      let newline_re = regex::Regex::new("\r?\n").unwrap();
+      self
+        .write_all(newline_re.replace_all(text, "\r\n").as_bytes())
+        .unwrap();
+    }
+  }
+
+  impl std::io::Write for WinPseudoConsole {
+    fn write(&mut self, buffer: &[u8]) -> std::io::Result<usize> {
+      unsafe {
+        let mut bytes_written = 0;
+        assert_win_success!(WriteFile(
+          self.stdin_write_handle.as_raw(),
+          buffer.as_ptr() as *const _,
+          buffer.len() as u32,
+          &mut bytes_written,
+          ptr::null_mut(),
+        ));
+        Ok(bytes_written as usize)
+      }
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+      Ok(())
+    }
+  }
+
+  struct WinHandle {
```

Comment:
In case you ever need to expand WinHandle functionality: try to follow the rust_std `OwnedHandle` implementation (which is unstable, but otherwise good). https://doc.rust-lang.org/beta/std/os/windows/io/struct.OwnedHandle.html